smarts/core/provider.py

```diff
@@ -18,13 +18,25 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 from dataclasses import dataclass, field
-from typing import List, Set
+from enum import IntFlag
+from typing import List, Optional, Set
 
 from .controllers import ActionSpaceType
 from .scenario import Scenario
 from .vehicle import VehicleState
 
 
+class ProviderRecoveryFlags(IntFlag):
+    NOT_REQUIRED = 0x00000000
+    """Not needed for the current step. Error causes skip."""
+    EPISODE_REQUIRED = 0x00000010
+    """Needed for the current episode. Results in episode ending."""
+    EXPERIMENT_REQUIRED = 0x00000100
+    """Needed for the experiment. Results in exception if an error is thrown."""
+    ATTEMPT_RECOVERY = 0x00001000
+    """Provider should attempt to recover from the exception or disconnection."""
```

Comment:
This seems to be orthogonal to the others.  For example, we might consider a Provider to be required for an episode _and_ want to attempt recovery, only ending the episode if recovery fails.

I'm guessing that's why you used non-overlapping bits for each of these enums.

I think it would be good to also indicate this with more of a comment on `ATTEMPT_RECOVERY` too.  Something like "Note: this can be combined with any of the above by creating a bitmask out of the enum values using `|`."